In [9]:
from tensorflow import keras
import tensorflow as tf
import numpy as np


keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[32, 32, 3]))
for _ in range(20):
    model.add(keras.layers.Dense(100,
                                 activation="elu",
                                 kernel_initializer="he_normal"))


In [10]:
model.add(keras.layers.Dense(10, activation="softmax"))

In [11]:
optimizer = keras.optimizers.Nadam(learning_rate=5e-5)
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=optimizer,
              metrics=["accuracy"])


In [4]:

(X_train_full, y_train_full), (X_test, y_test) = keras.datasets.cifar10.load_data()

X_train = X_train_full[5000:]
y_train = y_train_full[5000:]
X_valid = X_train_full[:5000]
y_valid = y_train_full[:5000]

In [14]:
import os
early_stopping_cb = keras.callbacks.EarlyStopping(patience=20)
model_checkpoint_cb = keras.callbacks.ModelCheckpoint("my_cifar10_model.h5", save_best_only=True)
run_index = 1 # increment every time you train the model
run_logdir = os.path.join(os.curdir, "my_cifar10_logs", "run_{:03d}".format(run_index))
tensorboard_cb = keras.callbacks.TensorBoard(run_logdir)
callbacks = [early_stopping_cb, model_checkpoint_cb, tensorboard_cb]

In [17]:
%load_ext tensorboard
%tensorboard --logdir=./my_cifar10_logs --port=6006

In [18]:
model.fit(X_train, y_train, epochs=100,
          validation_data=(X_valid, y_valid),
          callbacks=callbacks)

Epoch 1/100
1407/1407 [==============================] - 20s 12ms/step - loss: 4.0002 - accuracy: 0.1643 - val_loss: 2.1159 - val_accuracy: 0.2338
Epoch 2/100
1407/1407 [==============================] - 16s 11ms/step - loss: 2.0378 - accuracy: 0.2566 - val_loss: 2.1490 - val_accuracy: 0.2252
Epoch 3/100
1407/1407 [==============================] - 16s 11ms/step - loss: 1.9214 - accuracy: 0.2993 - val_loss: 1.9659 - val_accuracy: 0.2840
Epoch 4/100
1407/1407 [==============================] - 15s 11ms/step - loss: 1.8486 - accuracy: 0.3279 - val_loss: 1.9724 - val_accuracy: 0.3064
Epoch 5/100
1407/1407 [==============================] - 15s 11ms/step - loss: 1.7903 - accuracy: 0.3522 - val_loss: 1.7818 - val_accuracy: 0.3592
Epoch 6/100
1407/1407 [==============================] - 15s 11ms/step - loss: 1.7434 - accuracy: 0.3663 - val_loss: 1.7422 - val_accuracy: 0.3786
Epoch 7/100
1407/1407 [==============================] - 15s 10ms/step - loss: 1.7010 - accuracy: 0.3851 - val_loss: 1

In [19]:
model = keras.models.load_model("my_cifar10_model.h5")
model.evaluate(X_valid, y_valid)

157/157 [==============================] - 1s 2ms/step - loss: 1.4767 - accuracy: 0.4860


[1.4766956567764282, 0.4860000014305115]

In [5]:
from tensorflow import keras
import tensorflow as tf
import numpy as np
import os

keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[32, 32, 3]))
model.add(keras.layers.BatchNormalization())
for _ in range(20):
    model.add(keras.layers.Dense(100, kernel_initializer="he_normal"))
    model.add(keras.layers.BatchNormalization())
    model.add(keras.layers.Activation("elu"))
model.add(keras.layers.Dense(10, activation="softmax"))

optimizer = keras.optimizers.Nadam(learning_rate=5e-4)
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=optimizer,
              metrics=["accuracy"])

early_stopping_cb = keras.callbacks.EarlyStopping(patience=20)
model_checkpoint_cb = keras.callbacks.ModelCheckpoint("my_cifar10_bn_model.h5", save_best_only=True)
run_index = 1 # increment every time you train the model
run_logdir = os.path.join(os.curdir, "my_cifar10_logs", "run_bn_{:03d}".format(run_index))
tensorboard_cb = keras.callbacks.TensorBoard(run_logdir)
callbacks = [early_stopping_cb, model_checkpoint_cb, tensorboard_cb]

model.fit(X_train, y_train, epochs=100,
          validation_data=(X_valid, y_valid),
          callbacks=callbacks)

model = keras.models.load_model("my_cifar10_bn_model.h5")
model.evaluate(X_valid, y_valid)

Epoch 1/100
1407/1407 [==============================] - 35s 18ms/step - loss: 1.8390 - accuracy: 0.3428 - val_loss: 1.6808 - val_accuracy: 0.3984
Epoch 2/100
1407/1407 [==============================] - 22s 15ms/step - loss: 1.6690 - accuracy: 0.4045 - val_loss: 1.6186 - val_accuracy: 0.4276
Epoch 3/100
1407/1407 [==============================] - 21s 15ms/step - loss: 1.5990 - accuracy: 0.4312 - val_loss: 1.5566 - val_accuracy: 0.4374
Epoch 4/100
1407/1407 [==============================] - 22s 16ms/step - loss: 1.5494 - accuracy: 0.4490 - val_loss: 1.5179 - val_accuracy: 0.4614
Epoch 5/100
1407/1407 [==============================] - 21s 15ms/step - loss: 1.5100 - accuracy: 0.4653 - val_loss: 1.4386 - val_accuracy: 0.4904
Epoch 6/100
1407/1407 [==============================] - 21s 15ms/step - loss: 1.4693 - accuracy: 0.4794 - val_loss: 1.4183 - val_accuracy: 0.4960
Epoch 7/100
1407/1407 [==============================] - 22s 15ms/step - loss: 1.4355 - accuracy: 0.4896 - val_loss: 1

[1.3017864227294922, 0.5526000261306763]

<h3>Adding a <i>Batch Normalisation</i> Layer to improvise the model accuracy. Also, it tries to preserve a mean of 0 and standard deviation of 1. The model manages to achieve an accuracy of 55% which is much better than what the previous model achieved.</h3>

In [6]:
from tensorflow import keras
import tensorflow as tf
import numpy as np
import os

keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[32, 32, 3]))
model.add(keras.layers.BatchNormalization())
for _ in range(20):
    model.add(keras.layers.Dense(100, kernel_initializer="he_normal"))
    model.add(keras.layers.Activation("elu"))
    model.add(keras.layers.BatchNormalization())
model.add(keras.layers.Dense(10, activation="softmax"))

optimizer = keras.optimizers.Nadam(learning_rate=5e-4)
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=optimizer,
              metrics=["accuracy"])

early_stopping_cb = keras.callbacks.EarlyStopping(patience=20)
model_checkpoint_cb = keras.callbacks.ModelCheckpoint("my_cifar10_bn_model.h5", save_best_only=True)
run_index = 1 # increment every time you train the model
run_logdir = os.path.join(os.curdir, "my_cifar10_logs", "run_bn_{:03d}".format(run_index))
tensorboard_cb = keras.callbacks.TensorBoard(run_logdir)
callbacks = [early_stopping_cb, model_checkpoint_cb, tensorboard_cb]

model.fit(X_train, y_train, epochs=100,
          validation_data=(X_valid, y_valid),
          callbacks=callbacks)

model = keras.models.load_model("my_cifar10_bn_model.h5")
model.evaluate(X_valid, y_valid)

Epoch 1/100
1407/1407 [==============================] - 46s 26ms/step - loss: 1.9320 - accuracy: 0.3076 - val_loss: 1.7105 - val_accuracy: 0.3842
Epoch 2/100
1407/1407 [==============================] - 24s 17ms/step - loss: 1.7204 - accuracy: 0.3848 - val_loss: 1.6062 - val_accuracy: 0.4254
Epoch 3/100
1407/1407 [==============================] - 22s 16ms/step - loss: 1.6371 - accuracy: 0.4183 - val_loss: 1.5485 - val_accuracy: 0.437042 - ETA:  - ETA: 
Epoch 4/100
1407/1407 [==============================] - 24s 17ms/step - loss: 1.5831 - accuracy: 0.4371 - val_loss: 1.4933 - val_accuracy: 0.4644
Epoch 5/100
1407/1407 [==============================] - 24s 17ms/step - loss: 1.5356 - accuracy: 0.4574 - val_loss: 1.4592 - val_accuracy: 0.4784
Epoch 6/100
1407/1407 [==============================] - 25s 17ms/step - loss: 1.4993 - accuracy: 0.4686 - val_loss: 1.4431 - val_accuracy: 0.4912
Epoch 7/100
1407/1407 [==============================] - 23s 17ms/step - loss: 1.4608 - accuracy: 0.

[1.332292079925537, 0.5401999950408936]

<h3>Tried placing the <i>Batch Normalisation</i> Layer after the <i>ELU</i> activation function in the hidden layer, to see if the model reaches better accuracy. 

The model manages to achieve an accuracy of 54%; The BN layer slightly performs better when placed before the activation function in the hidden layers.</h3>

In [7]:
from tensorflow import keras
import tensorflow as tf
import numpy as np
import os

keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[32, 32, 3]))
for _ in range(20):
    model.add(keras.layers.Dense(100,
                                 kernel_initializer="lecun_normal",
                                 activation="selu"))
model.add(keras.layers.Dense(10, activation="softmax"))

optimizer = keras.optimizers.Nadam(learning_rate=7e-4)
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=optimizer,
              metrics=["accuracy"])

early_stopping_cb = keras.callbacks.EarlyStopping(patience=20)
model_checkpoint_cb = keras.callbacks.ModelCheckpoint("my_cifar10_selu_model.h5", save_best_only=True)
run_index = 1 # increment every time you train the model
run_logdir = os.path.join(os.curdir, "my_cifar10_logs", "run_selu_{:03d}".format(run_index))
tensorboard_cb = keras.callbacks.TensorBoard(run_logdir)
callbacks = [early_stopping_cb, model_checkpoint_cb, tensorboard_cb]

X_means = X_train.mean(axis=0)
X_stds = X_train.std(axis=0)
X_train_scaled = (X_train - X_means) / X_stds
X_valid_scaled = (X_valid - X_means) / X_stds
X_test_scaled = (X_test - X_means) / X_stds

model.fit(X_train_scaled, y_train, epochs=100,
          validation_data=(X_valid_scaled, y_valid),
          callbacks=callbacks)

model = keras.models.load_model("my_cifar10_selu_model.h5")
model.evaluate(X_valid_scaled, y_valid)

Epoch 1/100
1407/1407 [==============================] - 25s 15ms/step - loss: 1.9028 - accuracy: 0.3216 - val_loss: 1.8474 - val_accuracy: 0.3482
Epoch 2/100
1407/1407 [==============================] - 15s 11ms/step - loss: 1.6984 - accuracy: 0.3999 - val_loss: 1.7947 - val_accuracy: 0.3474
Epoch 3/100
1407/1407 [==============================] - 16s 11ms/step - loss: 1.6019 - accuracy: 0.4346 - val_loss: 1.7130 - val_accuracy: 0.3970
Epoch 4/100
1407/1407 [==============================] - 15s 11ms/step - loss: 1.5365 - accuracy: 0.4613 - val_loss: 1.6063 - val_accuracy: 0.4456
Epoch 5/100
1407/1407 [==============================] - 16s 11ms/step - loss: 1.4802 - accuracy: 0.4825 - val_loss: 1.5612 - val_accuracy: 0.4598
Epoch 6/100
1407/1407 [==============================] - 16s 11ms/step - loss: 1.4378 - accuracy: 0.4985 - val_loss: 1.5162 - val_accuracy: 0.4764
Epoch 7/100
1407/1407 [==============================] - 16s 11ms/step - loss: 1.3913 - accuracy: 0.5130 - val_loss: 1

[1.479697346687317, 0.4885999858379364]

In [8]:
 model = keras.models.load_model("my_cifar10_selu_model.h5")
model.evaluate(X_valid_scaled, y_valid)

157/157 [==============================] - 1s 2ms/step - loss: 1.4797 - accuracy: 0.4886


[1.479697346687317, 0.4885999858379364]

 
<h2>The model reaches 48.6% accuracy on the validation set using <i>SELU</i> Activation Function & Kernel_initialiser of <i>Lecun_Normal</i>.</h2>

In [9]:
keras.backend.clear_session()
tf.random.set_seed(42)
np.random.seed(42)

model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[32, 32, 3]))
for _ in range(20):
    model.add(keras.layers.Dense(100,
                                 kernel_initializer="lecun_normal",
                                 activation="selu"))

model.add(keras.layers.AlphaDropout(rate=0.1))
model.add(keras.layers.Dense(10, activation="softmax"))

optimizer = keras.optimizers.Nadam(learning_rate=5e-4)
model.compile(loss="sparse_categorical_crossentropy",
              optimizer=optimizer,
              metrics=["accuracy"])

early_stopping_cb = keras.callbacks.EarlyStopping(patience=20)
model_checkpoint_cb = keras.callbacks.ModelCheckpoint("my_cifar10_alpha_dropout_model.h5", save_best_only=True)
run_index = 1 # increment every time you train the model
run_logdir = os.path.join(os.curdir, "my_cifar10_logs", "run_alpha_dropout_{:03d}".format(run_index))
tensorboard_cb = keras.callbacks.TensorBoard(run_logdir)
callbacks = [early_stopping_cb, model_checkpoint_cb, tensorboard_cb]

X_means = X_train.mean(axis=0)
X_stds = X_train.std(axis=0)
X_train_scaled = (X_train - X_means) / X_stds
X_valid_scaled = (X_valid - X_means) / X_stds
X_test_scaled = (X_test - X_means) / X_stds

model.fit(X_train_scaled, y_train, epochs=100,
          validation_data=(X_valid_scaled, y_valid),
          callbacks=callbacks)

model = keras.models.load_model("my_cifar10_alpha_dropout_model.h5")
model.evaluate(X_valid_scaled, y_valid)

Epoch 1/100
1407/1407 [==============================] - 26s 16ms/step - loss: 1.8793 - accuracy: 0.3355 - val_loss: 1.7408 - val_accuracy: 0.3726
Epoch 2/100
1407/1407 [==============================] - 16s 11ms/step - loss: 1.6517 - accuracy: 0.4148 - val_loss: 1.7057 - val_accuracy: 0.3904
Epoch 3/100
1407/1407 [==============================] - 16s 11ms/step - loss: 1.5650 - accuracy: 0.4497 - val_loss: 1.6031 - val_accuracy: 0.4422
Epoch 4/100
1407/1407 [==============================] - 16s 11ms/step - loss: 1.5025 - accuracy: 0.4688 - val_loss: 1.5776 - val_accuracy: 0.4570
Epoch 5/100
1407/1407 [==============================] - 16s 11ms/step - loss: 1.4455 - accuracy: 0.4907 - val_loss: 1.7218 - val_accuracy: 0.4526
Epoch 6/100
1407/1407 [==============================] - 16s 11ms/step - loss: 1.4008 - accuracy: 0.5126 - val_loss: 1.5387 - val_accuracy: 0.4766
Epoch 7/100
1407/1407 [==============================] - 16s 11ms/step - loss: 1.3608 - accuracy: 0.5247 - val_loss: 1

[1.5122541189193726, 0.4887999892234802]

<h2> The model reaches an accuracy of 48.8% on the validation set using Alpha Dropout. </h2>